### PR TITLE
deprecate NIOEventLoopGroupProvider.createNew

### LINKED ISSUE #2142: ecosystem issue: EventLoopGroupProvider and the 'similar create or share ELG' pattern
The NIO ecosystem suffers from one composability issue: `EventLoopGroupProvider` and similar constructs. The issue is that taking `EventLoopGroupProvider` (or similar) means that a library can either own or not own an `EventLoopGroup`.

That doesn't sound too bad but unfortunately, it's very bad for the shutdown case.

A library (like say AsyncHTTPClient) would obviously want to have a method called `thing.shutdown() -> EventLoopFuture<Void>` to match the other functions. Unfortunately, that can't be done because the library may own the `EventLoopGroup`, therefore `shutdown()` has to may have to shut it down which then however means that there is no more `EventLoop` left to run the future on.

The current hack is to instead provide a `shutdown(group: DispatchGroup, _ completion: @escaping (Error?) -> Void)` method. But that's both ugly (doesn't fit the rest) and wasteful (needs to spin up a dispatch thread for no reason).

Unfortunately, this issue won't just magically be fixed by the universal adoption of Swift Concurrency either. Yes, it's possible to provide a `func shutdown() async throws -> Void` but that'll then bounce from a Concurrency executor to the EventLoop to a DispatchQueue thread back to some Concurrency executor. That's still not great.

Also, there's another issue if `EventLoop(Group)`s frequently go away: What do we do with people who caught a reference to an EL(G) and then (after the shutdown) `execute` stuff on it? Currently we print this weird warning that we'll crash in future releases. That's cool if EL(G)s are mostly global/created in main but it really doesn't work if they're frequently created/destroyed.

Also there's a question w.r.t. Custom Executors (once they actually arrive in Swift): Will they support this use case?

```
func myFunc() async throws -> Void {
    let httpClient = HTTPClient(eventLoopGroup: .createNew)
    // on some Concurrency executor
    do {
        let result = try await httpClient.get(...)
        // on a NIO EventLoop Custom Executor
        print(result)
        // we should be sure that this is actually legal to shut down the executor we're currently
        try await httpClient.shutdown() // this is also ugly because no async in defer
    } catch {
        try await httpClient.shutdown() // this is also ugly because no async in defer
    }
}
```

---

My recommendations are:
- deprecate `EventLoopGroupProvider`
- deprecate the pattern that an instance of a library (like `HTTPClient`) creates its own EL(G) threads

If we did deprecate `EventLoopGroupProvider` and the whole pattern, what should users do who don't already have an EL(G)? This will also become even more prevalent in the future where every user-facing API will be Swift Concurrency?

The only solution I can come up with is for either SwiftNIO itself or every library lazily creating an internal ELG that will _never_ be shut down. In other words: A global EL(G) singleton that gets created on first use, much like Dispatch or Swift Concurrency work. From a resource perspective and also alignment with Swift Concurrency it might actually make sense if SwiftNIO owned that singleton EL(G). That'd mean the new pattern to replace `EventLoopGroupProvider` would be `.shared(elg)` or `.global`.

The alternative to SwiftNIO owning one global ELG would be ofc for every library owning its own global ELG that could maybe just have 1 thread. I.e. AsyncHTTPClient would internally have a singleton ELG with just one thread.

What do you all think about this? None of this is fully baked but I'm really quite sure that the `EventLoopGroupProvider` (and similar) pattern is really bad.

### PR DESCRIPTION
**Pull request by weissi**
### Motivation:

`NIOEventLoopGroupProvider.createNew` was probably never a good idea because it creates shutdown issues for any library that uses it. Given that we now have singleton (#2471) `EventLoopGroup`s, we can solve this issue by just not having event loop group providers.

Users can just use `group: any EventLoopGroup` and optionally `group: any EventLoopGroup = MultiThreadedEventLoopGroup.singleton`.

### Modifications:

- deprecate `NIOEventLoopGroupProvider.createNew`
- soft-deprecate (document as deprecated but don't mark `@available(deprecated)`) `NIOEventLoopGroupProvider`

### Result:

- Libraries becomes easier to write and maintain.
- Fixes #2142

### CODE REVIEW

**Comment by weissi** [comment]
Must be merged _after_ #2471

**Comment by adam-fowler** [comment]
`group: any EventLoopGroup = MultiThreadedEventLoopGroup.singleton` is not a great alternative as it is no use if NIOPosix is not available. Also it would be good if https://github.com/apple/swift-nio-transport-services/pull/180 was also merged before this got merged.

Any library that supports both Posix and Network framework will probably end up having to implement a new version of `NIOEventLoopGroupProvider` which replaces `createNew` with `useSingleton` or something similar. It might be better if this type came from NIO though to provide consistency.

File: Sources/NIOCore/EventLoop.swift
Comment (Lukasa) [reply Sources/NIOCore/EventLoop.swift:1279]: I'm kinda with George here, I think it'd be better to do a multi-stage approach. Causing all user libraries to immediately get deprecation warnings without a cycle for maintainers to add new API surface seems like it's suboptimal.

**Comment by weissi** [comment]
> `group: any EventLoopGroup = MultiThreadedEventLoopGroup.singleton` is not a great alternative as it is no use if NIOPosix is not available.

That's a good point.

> Also it would be good if [apple/swift-nio-transport-services#180](https://github.com/apple/swift-nio-transport-services/pull/180) was also merged before this got merged.

Yes, I'll work on that rn.


> Any library that supports both Posix and Network framework will probably end up having to implement a new version of `NIOEventLoopGroupProvider` which replaces `createNew` with `useSingleton` or something similar. It might be better if this type came from NIO though to provide consistency.

I'd hope that libraries could use `eventLoopGroup: any EventLoopGroup? = nil` where `nil` means "default" and what that means can be chosen by the library itself. That way we also don't leak the exact value into the API. WDYT?

File: Sources/NIOCore/EventLoop.swift
Comment (weissi) [reply Sources/NIOCore/EventLoop.swift:1279]: > We could just deprecate the `createNew` case. End users can provide `.shared(MultiThreadedEventLoopGroup.singleton)` until libraries add new APIs which just take `any EventLoopGroup`.

Is the suggestion that libraries like Hummingbird and friends would add their own `HummingBird.defaultEventLoopGroup` or something and migrate from `groupProvider: EventLoopGroupProvider = .createNew` to `groupProvider: EventLoopGroupProvider = .shared(HummingBird.defaultEventLoopGroup)`?

At the same time they could add a `group: any EventLoopGroup? = nil`.

That would work for me. WDYT? @Lukasa / @glbrntt & @adam-fowler

Comment (FranzBusch) [reply Sources/NIOCore/EventLoop.swift:1279]: IMO, hummingbird and co just taking an EL is not great due to the relation ship of EL to bootstrap. I think we should just create that `swift-nio-universal-bootstrap` repo at some point where we expose a common way to bootstrap a client/server/datagram with a configuration that allows the user to choose `.automatic`, `.posix` or `.niots`. With the introduction of `swift-certificates` we have a way around the TLS configuration problem since we can just accept a `Certificate` and then serialise and deserialise.

If we would have that hummingbird and friends would accept a `UniversalServerBootstrapConfiguration` which would also include the EL configuration.

Comment (glbrntt) [reply Sources/NIOCore/EventLoop.swift:1279]: > > We could just deprecate the `createNew` case. End users can provide `.shared(MultiThreadedEventLoopGroup.singleton)` until libraries add new APIs which just take `any EventLoopGroup`.
> 
> Is the suggestion that libraries like Hummingbird and friends would add their own `HummingBird.defaultEventLoopGroup` or something and migrate from `groupProvider: EventLoopGroupProvider = .createNew` to `groupProvider: EventLoopGroupProvider = .shared(HummingBird.defaultEventLoopGroup)`?
> 
> At the same time they could add a `group: any EventLoopGroup? = nil`.
> 
> That would work for me. WDYT? @Lukasa / @glbrntt & @adam-fowler

Using the `.shared` would be the intermediate step so that end-users can avoid warnings while libraries put out new APIs.

I think the `group: any EventLoopGroup? = nil` where `nil` is "choose something sensible for me" is probably the right end goal. 

Why would libraries need to provide a `defaultEventLoopGroup`? Seems like just the `nil` case of above.

**Comment by adam-fowler** [comment]
> I'd hope that libraries could use `eventLoopGroup: any EventLoopGroup? = nil` where `nil` means "default" and what that means can be chosen by the library itself. That way we also don't leak the exact value into the API. WDYT?

That'd work I guess

File: Sources/NIOCore/EventLoop.swift
Comment (weissi) [reply Sources/NIOCore/EventLoop.swift:1279]: > Why would libraries need to provide a defaultEventLoopGroup? Seems like just the nil case of above.

@glbrntt Once we deprecate `.createNew`, what should Hummingbird change its API too? https://github.com/hummingbird-project/hummingbird/blob/c5dd619b1cd0a2683f0dd3d4b7a0bea12485060b/Sources/Hummingbird/Application.swift#L68C1-L73C1

Hummingbird can't change to `eventLoopGroupProvider: NIOEventLoopGroupProvider = .shared(MultiThreadedEventLoopGroup.singleton)` because that'll switch off NIOTS support by default. And they can't use `.createNew` (because deprecated).

The only interim step that's not deprecating the whole API (which I was _originally_ intending to do with the deprecation of `NIOEventLoopGroupProvider`) is if Hummingbird provided its own `.defaultEventLoopGroup` (which switches between posix & TS) and then goes for

`eventLoopGroupProvider: NIOEventLoopGroupProvider = .shared(...defaultEventLoopGroup)`

or am I missing smth?

Alongside that, HB probably wants to introduce a new `eventLoopGroup: any EventLoopGroup? = nil` API anyway.

Comment (glbrntt) [reply Sources/NIOCore/EventLoop.swift:1279]: > > Why would libraries need to provide a defaultEventLoopGroup? Seems like just the nil case of above.
> 
> @glbrntt Once we deprecate `.createNew`, what should Hummingbird change its API too? https://github.com/hummingbird-project/hummingbird/blob/c5dd619b1cd0a2683f0dd3d4b7a0bea12485060b/Sources/Hummingbird/Application.swift#L68C1-L73C1
> 
> Hummingbird can't change to `eventLoopGroupProvider: NIOEventLoopGroupProvider = .shared(MultiThreadedEventLoopGroup.singleton)` because that'll switch off NIOTS support by default. And they can't use `.createNew` (because deprecated).
> 
> The only interim step that's not deprecating the whole API (which I was _originally_ intending to do with the deprecation of `NIOEventLoopGroupProvider`) is if Hummingbird provided its own `.defaultEventLoopGroup` (which switches between posix & TS) and then goes for
> 
> `eventLoopGroupProvider: NIOEventLoopGroupProvider = .shared(...defaultEventLoopGroup)`
> 
> or am I missing smth?

I _think_ what you're missing is that the intermediate deprecation is to help _end users_, i.e. people using Hummingbird et. al., but not Hummingbird itself. 

What we should try to avoid is:
1. NIO is released, deprecating the provider
2. User updates and has deprecation warnings but no way to avoid them
3. Hummingbird et. al. get updated to provide an alternative API
4. User can move to alternative API

We're fine if 2 and 3 are reversed but we can't control that, so I suggest we do the following:
1. NIO is released, deprecating the `.createNew` case
2. User updates and has deprecation warning but can either live with the warning _or_ use an appropriate singleton in the `.shared` case
3. Hummingbird et. al. get updated to provide an alternative API and deprecate their API using the provider
5. User can move to new API
 
> Alongside that, HB probably wants to introduce a new `eventLoopGroup: any EventLoopGroup? = nil` API anyway.

The new API should be this ^

Comment (adam-fowler) [reply Sources/NIOCore/EventLoop.swift:1279]: > Once we deprecate .createNew, what should Hummingbird change its API too

I guess Hummingbird would deprecate the initialiser using `NIOEventLoopGroupProvider` and add a new version that took `(any EventLoopGroup)?`. Although the default parameters might complicate the issue.

**Review by 1proprogrammerchant — APPROVED** [review]

**Comment by weissi** [comment]
friendly ping @Lukasa

File: Sources/NIOCore/EventLoop.swift
Comment (Lukasa) [reply Sources/NIOCore/EventLoop.swift:1252]: @weissi Do you have any objections to @glbrntt's proposal above?

Comment (weissi) [reply Sources/NIOCore/EventLoop.swift:1252]: @Lukasa Yes, deprecating `.createNew` to start with works for me!

Comment (weissi) [reply Sources/NIOCore/EventLoop.swift:1252]: I think I did the same in AHC: https://github.com/swift-server/async-http-client/blob/bdaa3b18358b60e268afe82e8feb16418874f41b/Sources/AsyncHTTPClient/HTTPClient.swift#L956-L966

Comment (weissi) [reply Sources/NIOCore/EventLoop.swift]: done

**Comment by weissi** [comment]
@Lukasa / @glbrntt addressed, ptal

File: Sources/NIOCore/EventLoop.swift
Comment (glbrntt) [inline Sources/NIOCore/EventLoop.swift:1272]:
```suggestion
            Please use `.shared(existingGroup)` or use the singleton via
```

Comment (glbrntt) [inline Sources/NIOCore/EventLoop.swift:1272]: Do you actually want a multiline string here?

File: Sources/NIOCore/EventLoop.swift
Comment (glbrntt) [inline Sources/NIOCore/EventLoop.swift:1256]:
```suggestion
/// singleton group instead, consider group: any EventLoopGroup = MultiThreadedEventLoopGroup.singleton` or
```

File: Sources/NIOCore/EventLoop.swift
Comment (weissi) [reply Sources/NIOCore/EventLoop.swift]: ta

File: Sources/NIOCore/EventLoop.swift
Comment (weissi) [reply Sources/NIOCore/EventLoop.swift]: ta

File: Sources/NIOCore/EventLoop.swift
Comment (weissi) [reply Sources/NIOCore/EventLoop.swift]: Yes, didn't fit in one line. Is that bad?

Comment (weissi) [reply Sources/NIOCore/EventLoop.swift]: oh sorry, didn't want the `\n`,, fixing

Comment (weissi) [reply Sources/NIOCore/EventLoop.swift]: thank you, fixed

**Pull request merged**
